### PR TITLE
[front] - feature: implement user and agent usage reporting

### LINKED ISSUE #5949: [Workspace Metrics] - New queries

### PR DESCRIPTION
**Pull request by JulesBelveze**
## Description
This PR introduces new queries that will be added to the reporting for workspace analytics. 

The main changes include:
- Adding new SQL queries to fetch detailed usage data for users, agents, and messages within specified time periods.
- Implementing functions to generate CSV reports for message, user, builder, and agent usage data.

References:
- https://github.com/orgs/dust-tt/projects/3/views/1?pane=issue&itemId=67809458

## Risk
None for now as the new queries are not being used

## Deploy Plan
None

### CODE REVIEW

File: front/lib/workspace_usage.ts [hunk 1]
```diff
@@ -104,6 +150,301 @@ export async function unsafeGetUsageData(
...
+      replacements: {
+        startDate: format(startDate, "yyyy-MM-dd'T'00:00:00"), // Use first day of start month
+        endDate: format(endDate, "yyyy-MM-dd'T'23:59:59"), // Use last day of end month
+      },
+    }
+  );
+  if (!mentions.length) {
+    return "No data available for the selected period.";
+  }
+  const csvHeader = Object.keys(mentions[0]).join(",") + "\n";
```
Comment (philipperolet) [inline front/lib/workspace_usage.ts:440]: would dry the 3 usages of this

File: front/lib/workspace_usage.ts [hunk 2]
```diff
@@ -104,6 +150,301 @@ export async function unsafeGetUsageData(
...
+      a."createdAt" BETWEEN :startDate AND :endDate
+      AND ac."workspaceId" = ${wId}
+      AND ac."status" = 'active'
+      AND ac."scope" != 'private'
+    GROUP BY
+      ac."name",
+      ac."description",
+      ac."scope"
+    ORDER BY
+      "Messages Last 30 Days" DESC;
```
Comment (philipperolet) [inline front/lib/workspace_usage.ts:427]:
```suggestion
      "Messages Last 30 Days" DESC;
```
Does that work on sequelize? or is it a metabase thing?

File: front/lib/workspace_usage.ts [hunk 3]
```diff
@@ -104,6 +150,301 @@ export async function unsafeGetUsageData(
...
+        startDate: format(startDate, "yyyy-MM-dd'T'00:00:00"), // Use first day of start month
+        endDate: format(endDate, "yyyy-MM-dd'T'23:59:59"), // Use last day of end month
+      },
+    }
+  );
+  if (!mentions.length) {
+    return "No data available for the selected period.";
+  }
+  const csvHeader = Object.keys(mentions[0]).join(",") + "\n";
+  const csvContent = mentions
```
Comment (philipperolet) [inline front/lib/workspace_usage.ts:441]: would dry the 3 csv generation snippets

File: front/lib/workspace_usage.ts (hunk 2 above)
Comment (philipperolet) [reply front/lib/workspace_usage.ts:427]: if metabase, that won't work here; if SQL well... TIL

File: front/lib/workspace_usage.ts [hunk 4]
```diff
@@ -104,6 +150,301 @@ export async function unsafeGetUsageData(
...
+  const workspace = await Workspace.findOne({
+    where: { sId: workspaceId },
+  });
+  if (!workspace) {
+    throw new Error(`Workspace not found for sId: ${workspaceId}`);
+  }
+  const wId = workspace.id;
+  const mentions = await frontSequelize.query<AgentUsageQueryResult>(
+    `
+    SELECT
```
Comment (philipperolet) [inline front/lib/workspace_usage.ts:397]: nit: IMHO big queries like that make the whole file harder to read (+ psychological blow when seeing it :) )

would suggest creating separate .sql files for those queries; this way @thib-martin could even edit them directly (and make a comment as to what they are)
and we load them as top level constant
asking another eng for their opinion here e.g. @flvndvd

File: front/lib/workspace_usage.ts [hunk 5]
```diff
@@ -104,6 +150,301 @@ export async function unsafeGetUsageData(
...
+  const workspace = await Workspace.findOne({
+    where: { sId: workspaceId },
+  });
+  if (!workspace) {
+    throw new Error(`Workspace not found for sId: ${workspaceId}`);
+  }
+  const wId = workspace.id;
+  const results = await frontSequelize.query<MessageUsageQueryResult>(
+    `
+      SELECT
```
Comment (philipperolet) [inline front/lib/workspace_usage.ts:167]: see comment below for big sql queries

File: front/lib/workspace_usage.ts [hunk 6]
```diff
@@ -104,6 +150,301 @@ export async function unsafeGetUsageData(
   return csvHeader + csvContent;
 }
 
+export async function unsafeGetMessageUsageData(
```
Comment (philipperolet) [inline front/lib/workspace_usage.ts:153]: noob question: why unsafe? (maybe a com would help?)

**Review by philipperolet — COMMENTED** [review]
Thanks! Left a few coms. 2 things to check in addition:
- did you test it and it worked? (front-edge to have real data)
- will this be called on demand by admins only?
Asking to make sure that it's ok if the query is slow

File: front/lib/workspace_usage.ts (hunk 2 above)
Comment (JulesBelveze) [reply front/lib/workspace_usage.ts:427]: It indeed seems to work properly with Sequelize 😇

File: front/lib/workspace_usage.ts (hunk 6 above)
Comment (JulesBelveze) [reply front/lib/workspace_usage.ts:153]: No reason tbh, just got the naming pattern of the already existing function. Removing the "unsafe" prefix 👍🏼

File: front/lib/workspace_usage.ts (hunk 4 above)
Comment (JulesBelveze) [reply front/lib/workspace_usage.ts:397]: 💯 Sounds good to me!

**Comment by JulesBelveze** [comment]
> did you test it and it worked? (front-edge to have real data)
will this be called on demand by admins only?

@philipperolet I tested it on front-edge and it worked as expected, even though quite slow. But it will only be called on demand (basically replacing the "Download Activity Data" that we currently have

**Review by philipperolet — APPROVED** [review]

**Pull request merged**